ports/rp2/mpnetworkport.c

```diff
@@ -77,11 +77,64 @@ void cyw43_irq_init(void) {
     NVIC_SetPriority(PendSV_IRQn, IRQ_PRI_PENDSV);
 }
 
+void cyw43_irq_deinit(void) {
+    gpio_remove_raw_irq_handler(CYW43_PIN_WL_HOST_WAKE, gpio_irq_handler);
+}
+
 void cyw43_post_poll_hook(void) {
     cyw43_has_pending = 0;
     gpio_set_irq_enabled(CYW43_PIN_WL_HOST_WAKE, CYW43_IRQ_LEVEL, true);
 }
 
+#if CYW43_PIN_WL_DYNAMIC
+// Defined in cyw43_bus_pio_spi.c
+extern int cyw43_set_pins_wl(uint pins[CYW43_PIN_INDEX_WL_COUNT]);
+
+mp_obj_t network_cyw43_set_pins_wl(size_t n_args, const mp_obj_t *args) {
+    if (n_args == 6) {
+        uint pins[CYW43_PIN_INDEX_WL_COUNT] = {
+            // REG_ON, OUT, IN, WAKE, CLOCK, CS
+            mp_obj_get_int(args[0]), //CYW43_DEFAULT_PIN_WL_REG_ON,
+            mp_obj_get_int(args[1]), //CYW43_DEFAULT_PIN_WL_DATA_OUT,
+            mp_obj_get_int(args[2]), //CYW43_DEFAULT_PIN_WL_DATA_IN,
+            mp_obj_get_int(args[3]), //CYW43_DEFAULT_PIN_WL_HOST_WAKE,
+            mp_obj_get_int(args[4]), //CYW43_DEFAULT_PIN_WL_CLOCK,
+            mp_obj_get_int(args[5]), //CYW43_DEFAULT_PIN_WL_CS
+        };
+        cyw43_irq_deinit();
+        cyw43_set_pins_wl(pins);
+        cyw43_irq_init();
+    } else if (n_args == 4) {
+        uint pins[CYW43_PIN_INDEX_WL_COUNT] = {
+            // REG_ON, IO, CLOCK, CS
+            // Data Out, Data In and Host Wake are the same pin on Pico
+            mp_obj_get_int(args[0]), //CYW43_DEFAULT_PIN_WL_REG_ON,
+            mp_obj_get_int(args[1]), //CYW43_DEFAULT_PIN_WL_DATA_OUT,
+            mp_obj_get_int(args[1]), //CYW43_DEFAULT_PIN_WL_DATA_IN,
+            mp_obj_get_int(args[1]), //CYW43_DEFAULT_PIN_WL_HOST_WAKE,
+            mp_obj_get_int(args[2]), //CYW43_DEFAULT_PIN_WL_CLOCK,
+            mp_obj_get_int(args[3]), //CYW43_DEFAULT_PIN_WL_CS
+        };
+        cyw43_irq_deinit();
+        cyw43_set_pins_wl(pins);
+        cyw43_irq_init();
+    }
+    return mp_const_none;
+}
+MP_DEFINE_CONST_FUN_OBJ_VAR_BETWEEN(network_cyw43_set_pins_wl_obj, 4, 6, network_cyw43_set_pins_wl);
```

Comment:
Not sure that this is the best way to expose this configuration as a Python-level API.  For this kind of thing the usual approach would be to use the WLAN constructor, eg:
```py
wl = network.WLAN(reg_on=Pin(4), data_out=Pin(5), ..., freq=1_000_000)
```
And then have sensible defaults so it works on Pico 2 W with just `wl = network.WLAN()`.

Reply:
That would be neater. But these parameters are pico specific and network_cyw43_make_new is "common". I'll see if I can override this.

Reply:
I've had a go at doing this. I had to make changes to extmod/network_cyw43.c to make network_cyw43_make_new configurable.  So now you can do something like this...

wlan = network.WLAN(network.STA_IF, pin_on=2, pin_out=4, pin_in=4, pin_wake=4, pin_clock=5, pin_cs=3, div=3, div_frac=0)

Reply:
I like this, it's *much* nicer than my solution.

How does this play with the early pin setup that happens in `main.c` before any user code is run? It would be nice to decouple CYW43 from `main.c` altogether if possible, it doesn't feel like it should get any special treatment as a network interface, but it's doing all of this up front:

https://github.com/micropython/micropython/blob/18c9abc32930728d0403d8c13994adc3cabba8db/ports/rp2/main.c#L135-L155

Could this be moved to your custom `network_cyw43_make_new` somehow?

Reply:
That's probably there for "historical" reasons. I'll have to check if we can remove it or put it somewhere better.
I think I should add "pin_dat" that sets pin_in, pin_out and pin_wake in one go, as they should really always be the same.

Reply:
> That's probably there for "historical" reasons.

Oof. Programming terms to make your blood run cold!

> I think I should add "pin_dat" that sets pin_in, pin_out and pin_wake in one go, as they should really always be the same.

I don't foresee any situation where they wouldn't be, at least for Pimoroni boards or breakouts. Is it even possible for them not to be with the PIO driver? Anyway - yes please, the shorter the better since we'll definitely drop my hacky `set_pins` function in favour of this.

Reply:
> Is it even possible for them not to be with the PIO driver?

It looks like it might work, but it's untested so who knows!

Reply:
I added pin_dat...

wlan = network.WLAN(network.STA_IF, pin_on=2, pin_cs=3, pin_dat=4, pin_clock=5, div_int=3, div_frac=0)

And moved the cyw43 init code from main.